edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceBridge.cs

```diff
@@ -0,0 +1,320 @@
+                {
+                    return false;
+                }
+
+                return this.subscriptions.Contains(subscription);
+            }
+        }
+
+        internal async Task<ITry<ICloudProxy>> TryRetrieveCloudProxyAsync()
+        {
+            var newTask = false;
+            Task<ITry<ICloudProxy>> task;
+            using (SyncLock.Lock(this.stateLock, OperationTimeout))
+            {
+                // if existing cloudProxy is active, reuse it
+                if (this.cloudProxy?.IsActive ?? false)
+                {
+                    return Try.Success(this.cloudProxy);
+                }
+
+                // reuse existing createCloudProxyTask
+                if (this.createCloudProxyTask != null)
+                {
+                    task = this.createCloudProxyTask;
+                }
+                else
+                {
+                    // if no createCloudProxyTask, create new createCloudProxyTask
+                    task = this.TryCreateCloudProxyAsync();
+                    this.createCloudProxyTask = task;
+                    newTask = true;
+                }
+            }
+
+            var result = await task;
+
+            using (SyncLock.Lock(this.stateLock, OperationTimeout))
+            {
+                if (!newTask)
+                {
+                    // if reuse existing task, just return result
+                    return result;
+                }
+
+                // double check if task is the same as existing createCloudProxyTask, return the result
+                if (task == this.createCloudProxyTask)
+                {
+                    this.Debugging($"Create cloud proxy for {this.Identity} result={result.Success}.");
+                    this.createCloudProxyTask = null;
+                    if (result.Success)
+                    {
+                        this.cloudProxy = result.Value;
+                    }
+
+                    return result;
+                }
+            }
+
+            // task is different from existing createCloudProxyTask, it means task was cancelled and we need to close created ICloudProxy.
+            // Otherwise it will leak a DeviceClient instance and play pingpong
+            if (result.Success)
+            {
+                await result.Value.CloseAsync();
+            }
+
+            return Try.Failure<ICloudProxy>(new OperationCanceledException("Operation is cancelled."));
+        }
+
+        internal async Task<Option<ICloudProxy>> RetrieveCloudProxyAsync() => (await this.TryRetrieveCloudProxyAsync()).ToOption();
+
+        async Task<ITry<ICloudProxy>> TryCreateCloudProxyAsync()
+        {
+            Action<string, CloudConnectionStatus> cloudConnectionStatusChangedHandler;
+            using (SyncLock.Lock(this.stateLock, OperationTimeout))
+            {
+                var id = Guid.NewGuid().ToString();
+                this.cloudConnectionStatusChangedHandlerId = id;
```

Comment:
>this.cloudConnectionStatusChangedHandlerId = id; [](start = 16, length = 48)

is this to detect reentrancy so that only status events for the newest CloudProxy are handled?  If so let's add a hefty comment, it took me a while to parse through the usage of this ID.

Reply:
Will do